oracle_fdw.c

```diff
@@ -732,6 +734,69 @@ _PG_init(void)
 	on_proc_exit(&exitHook, PointerGetDatum(NULL));
 }
 
+/*
+ * Deparse LIMIT clause translated for ORACLE into FETCH FIRST N ROWS ONLY
+ * If OFFSET is set then do not append clause at all, the limit +offset
+ * will be applied localy.
+ */
+char *
+deparseLimit(PlannerInfo *root, struct OracleFdwState *fdwState, RelOptInfo *baserel)
+{
+	StringInfoData limit_clause;
+
+	initStringInfo(&limit_clause);
+
+	if (root->parse->limitCount != NULL && root->parse->limitOffset == NULL)
+	{
+		if (IsA(root->parse->limitCount, Const) && !((Const *) root->parse->limitCount)->constisnull)
+		{
+			int limit_at_end = false;
+
+			/*
+			 * we must be sure that there is no ORDER BY clause that
+			 * appears after the LIMIT clause otherwise we will have
+			 * wrong results.
+			 */
+			if (root->parse->sortClause != NULL)
+			{
+				ListCell   *lc;
+				foreach (lc, root->processed_tlist)
+				{
+					Node *ptl = (Node *) lfirst(lc);
+					if (IsA(ptl, TargetEntry))
+					{
+						TargetEntry *tge = (TargetEntry *) ptl;
+						/* Aggregate function are pushed down so it is safe to continue */
```

Comment:
Aggregate functions are *not* pushed down, and indeed the last test case shows that the patch is doing the wrong thing.
If there is a `GROUP BY` clause, you cannot push down the `LIMIT`, because `LIMIT` is calculated after `GROUP BY`.